Annotation method return type resolved as property for java sources.

When processing Java source for an annotation like:

```
@interface MyAnnotation {
    boolean booleanParam();
}
```

I'm hitting a `ClassCastException` when trying to get the return type of the booleanParam() method.

```
Caused by: java.lang.ClassCastException: class org.jetbrains.kotlin.load.java.descriptors.JavaPropertyDescriptor cannot be cast to class org.jetbrains.kotlin.descriptors.FunctionDescriptor (org.jetbrains.kotlin.load.java.descriptors.JavaPropertyDescriptor and org.jetbrains.kotlin.descriptors.FunctionDescriptor are in unnamed module of loader 'app')
    at com.google.devtools.ksp.processing.impl.ResolverImpl.resolveFunctionDeclaration(ResolverImpl.kt:640)
    at com.google.devtools.ksp.symbol.impl.java.KSFunctionDeclarationJavaImpl.findOverridee(KSFunctionDeclarationJavaImpl.kt:55)
    at androidx.room.compiler.processing.ksp.KSFunctionExtKt.getOriginatingReference(KSFunctionExt.kt:57)
    at androidx.room.compiler.processing.ksp.KSFunctionExtKt.returnKspType(KSFunctionExt.kt:35)
    at androidx.room.compiler.processing.ksp.KspMethodElement$KspNormalMethodElement$returnType$2.invoke(KspMethodElement.kt:165)
    at androidx.room.compiler.processing.ksp.KspMethodElement$KspNormalMethodElement$returnType$2.invoke(KspMethodElement.kt:164)
    at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
    at androidx.room.compiler.processing.ksp.KspMethodElement$KspNormalMethodElement.getReturnType(KspMethodElement.kt:164)
    at androidx.room.compiler.processing.XAnnotationValueTest$tmp$1.invoke(XAnnotationValueTest.kt:762)
    at androidx.room.compiler.processing.XAnnotationValueTest$tmp$1.invoke(XAnnotationValueTest.kt:724)
```

Also note that it's only when processing the annotation as a source -- it works as expected when the using the precompiled class

--- a/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt
+++ b/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt
@@ -352,7 +352,11 @@ class ResolverImpl(
     internal fun mapToJvmSignatureInternal(declaration: KSDeclaration): String? = when (declaration) {
         is KSClassDeclaration -> resolveClassDeclaration(declaration)?.let { typeMapper.mapType(it).descriptor }
         is KSFunctionDeclaration -> resolveFunctionDeclaration(declaration)?.let {
-            typeMapper.mapAsmMethod(it).descriptor
+            when (it) {
+                is FunctionDescriptor -> typeMapper.mapAsmMethod(it).descriptor
+                is PropertyDescriptor -> typeMapper.mapFieldSignature(it.type, it) ?: typeMapper.mapType(it).descriptor
+                else -> throw IllegalStateException("Unexpected descriptor type for declaration: $declaration")
+            }
         }
         is KSPropertyDeclaration -> resolvePropertyDeclaration(declaration)?.let {
             typeMapper.mapFieldSignature(it.type, it) ?: typeMapper.mapType(it).descriptor
@@ -414,7 +418,8 @@ class ResolverImpl(
         incrementalContext.recordLookupForDeclaration(candidate)
         val originalDescriptor = when (original) {
             is KSPropertyDeclaration -> resolvePropertyDeclaration(original)
-            is KSFunctionDeclaration -> resolveFunctionDeclaration(original)?.propertyIfAccessor
+            is KSFunctionDeclaration ->
+                resolveFunctionDeclaration(original).safeAs<FunctionDescriptor>()?.propertyIfAccessor
             else -> return false
         }
 
@@ -557,7 +562,7 @@ class ResolverImpl(
         } as ClassDescriptor?
     }
 
-    fun resolveFunctionDeclaration(function: KSFunctionDeclaration): FunctionDescriptor? {
+    fun resolveFunctionDeclaration(function: KSFunctionDeclaration): CallableDescriptor? {
         return when (function) {
             is KSFunctionDeclarationImpl -> resolveDeclaration(function.ktFunction)
             is KSFunctionDeclarationDescriptorImpl -> function.descriptor
@@ -581,7 +586,7 @@ class ResolverImpl(
                 resolved?.unsubstitutedPrimaryConstructor ?: resolved?.constructors?.singleOrNull()
             }
             else -> throw IllegalStateException("unexpected class: ${function.javaClass}")
-        } as FunctionDescriptor?
+        } as? CallableDescriptor
     }
 
     fun resolvePropertyDeclaration(property: KSPropertyDeclaration): PropertyDescriptor? {
@@ -807,9 +812,9 @@ class ResolverImpl(
     override fun getJvmName(declaration: KSFunctionDeclaration): String? {
         // function names might be mangled if they receive inline class parameters or they are internal
         val descriptor = resolveFunctionDeclaration(declaration)
-        return descriptor?.let {
+        return descriptor.safeAs<FunctionDescriptor>()?.let {
             // KotlinTypeMapper.mapSignature always uses OwnerKind.IMPLEMENTATION
-            typeMapper.mapFunctionName(descriptor, OwnerKind.IMPLEMENTATION)
+            typeMapper.mapFunctionName(it, OwnerKind.IMPLEMENTATION)
         }
     }
 

--- a/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/java/KSFunctionDeclarationJavaImpl.kt
+++ b/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/java/KSFunctionDeclarationJavaImpl.kt
@@ -28,6 +28,8 @@ import com.google.devtools.ksp.toKSModifiers
 import com.intellij.lang.jvm.JvmModifier
 import com.intellij.psi.PsiJavaFile
 import com.intellij.psi.PsiMethod
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
 class KSFunctionDeclarationJavaImpl private constructor(val psi: PsiMethod) :
     KSFunctionDeclaration,
@@ -53,7 +55,7 @@ class KSFunctionDeclarationJavaImpl private constructor(val psi: PsiMethod) :
 
     override fun findOverridee(): KSDeclaration? {
         val descriptor = ResolverImpl.instance.resolveFunctionDeclaration(this)
-        return descriptor?.findClosestOverridee()?.toKSDeclaration()
+        return descriptor.safeAs<FunctionDescriptor>()?.findClosestOverridee()?.toKSDeclaration()
     }
 
     override val declarations: Sequence<KSDeclaration> = emptySequence()

--- a/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/kotlin/KSFunctionDeclarationImpl.kt
+++ b/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/kotlin/KSFunctionDeclarationImpl.kt
@@ -32,6 +32,7 @@ import org.jetbrains.kotlin.psi.KtFunctionLiteral
 import org.jetbrains.kotlin.psi.KtNamedFunction
 import org.jetbrains.kotlin.psi.psiUtil.startOffset
 import org.jetbrains.kotlin.resolve.calls.inference.returnTypeOrNothing
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
 class KSFunctionDeclarationImpl private constructor(val ktFunction: KtFunction) :
     KSFunctionDeclaration,
@@ -43,7 +44,7 @@ class KSFunctionDeclarationImpl private constructor(val ktFunction: KtFunction) 
 
     override fun findOverridee(): KSDeclaration? {
         val descriptor = ResolverImpl.instance.resolveFunctionDeclaration(this)
-        return descriptor?.findClosestOverridee()?.toKSDeclaration()
+        return descriptor.safeAs<FunctionDescriptor>()?.findClosestOverridee()?.toKSDeclaration()
     }
 
     override val simpleName: KSName by lazy {

--- a/compiler-plugin/testData/api/overridee.kt
+++ b/compiler-plugin/testData/api/overridee.kt
@@ -85,6 +85,10 @@
 // JavaAccessorImpl.getX() -> KtInterfaceWithProperty.x
 // JavaAccessorImpl.getY() -> KtInterfaceWithProperty.y
 // JavaAccessorImpl.setY(value:Int) -> KtInterfaceWithProperty.y
+// JavaAnno:
+// JavaAnno.intParam() -> null
+// JavaAnnos:
+// JavaAnnos.value() -> null
 // END
 // MODULE: lib
 // FILE: lib.kt
@@ -294,4 +298,15 @@ class JavaAccessorImpl implements KtInterfaceWithProperty {
     }
     public void setY(int value) {
     }
+}
+
+// FILE: JavaAnno.java
+@java.lang.annotation.Repeatable(JavaAnnos.class)
+public @interface JavaAnno {
+    int intParam();
+}
+
+// FILE: JavaAnnos.java
+public @interface JavaAnnos {
+    JavaAnno[] value();
 }

--- a/compiler-plugin/testData/api/signatureMapper.kt
+++ b/compiler-plugin/testData/api/signatureMapper.kt
@@ -25,6 +25,9 @@
 // getVoid: ()Ljava/lang/Void;
 // LJavaClass;
 // <init>: ()V
+// LJavaAnno;
+// intParam: I
+// <init>: (I)V
 // END
 
 // FILE: Cls.kt
@@ -42,4 +45,9 @@ interface JavaIntefaceWithVoid {
 // FILE: JavaClass.java
 class JavaClass {
     JavaClass() {}
+}
+
+// FILE: JavaAnno.java
+@interface JavaAnno {
+    int intParam();
 }

--- a/test-utils/src/main/kotlin/com/google/devtools/ksp/processor/MapSignatureProcessor.kt
+++ b/test-utils/src/main/kotlin/com/google/devtools/ksp/processor/MapSignatureProcessor.kt
@@ -31,7 +31,7 @@ class MapSignatureProcessor : AbstractTestProcessor() {
     }
 
     override fun process(resolver: Resolver): List<KSAnnotated> {
-        listOf("Cls", "JavaIntefaceWithVoid", "JavaClass")
+        listOf("Cls", "JavaIntefaceWithVoid", "JavaClass", "JavaAnno")
             .map { className ->
                 resolver.getClassDeclarationByName(className)!!
             }.forEach { subject ->

--- a/test-utils/src/main/kotlin/com/google/devtools/ksp/processor/OverrideeProcessor.kt
+++ b/test-utils/src/main/kotlin/com/google/devtools/ksp/processor/OverrideeProcessor.kt
@@ -40,6 +40,8 @@ class OverrideeProcessor : AbstractTestProcessor() {
         logSubject(resolver, "OverrideOrder1")
         logSubject(resolver, "OverrideOrder2")
         logSubject(resolver, "JavaAccessorImpl")
+        logSubject(resolver, "JavaAnno")
+        logSubject(resolver, "JavaAnnos")
         return emptyList()
     }
 

